src/napari_threedee/data_models/surfaces.py

```diff
@@ -0,0 +1,176 @@
+import os
+from typing import List
+
+import napari
+import numpy as np
+import pandas as pd
+import zarr
+from pydantic import BaseModel, validator
+
+from napari_threedee.annotators.base import N3dDataModel
+from napari_threedee.annotators.constants import N3D_METADATA_KEY, ANNOTATION_TYPE_KEY
+from napari_threedee.annotators.surfaces.constants import LEVEL_ID_FEATURES_KEY, \
+    SURFACE_ID_FEATURES_KEY, SURFACE_ANNOTATION_TYPE_KEY, COLOR_CYCLE
+from napari_threedee.annotators.surfaces.validation import validate_layer, \
+    validate_n3d_zarr
+
+
+class N3dSurface(BaseModel):
+    data: List[np.ndarray]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    @property
+    def ndim(self) -> int:
+        return self.data[0].shape[-1]
+
+    @validator('data', pre=True)
+    def ensure_float32_ndarray_list(cls, value):
+        data = [
+            np.asarray(arr, dtype=np.float32)
+            for arr
+            in value
+        ]
+        return data
+
+    @property
+    def n_points(self) -> int:
+        return sum(len(arr) for arr in self.data)
+
+    @property
+    def n_levels(self) -> int:
+        return len(self.data)
+
+    @property
+    def points(self) -> np.ndarray:
+        return np.concatenate([level for level in self.data])
+
+    @property
+    def level_ids(self) -> np.ndarray:
+        level_ids = [
+            [idx] * len(level)
+            for idx, level
+            in enumerate(self.data)
+        ]
+        return np.concatenate(level_ids)
+
+    def __iter__(self) -> np.ndarray:
+        yield from self.data
+
+
+class N3dSurfaces(N3dDataModel):
+    data: List[N3dSurface]
```

Comment:
I think we could discuss this separately, but maybe the collections could be some sort of evented collection (e.g., `EventedList` or `EventedSet`). I'm not sure if it actually has any advantages, but just a thought after our annoyance with the selected `Points` yesterday. I think we can think about this later though.

Reply:
I made a conscious choice to make the models immutable so that people don't get the impression that they can modify the model and update the annotator. The rationale is that the napari layer model is the single source of truth for the annotator, the layer is mutable in the way layers usually are but must conform to a certain spec - producing a layer which exactly matches the spec is hairy, these objects are an intermediate which make it easy to do IO and produce layers matching the spec but they are not the data model for the annotator itself

I should write some docs for this :)

Reply:
Good point. I forgot about that. Totally makes sense. I really like the fact that they are immutable.